opentracing/trace_context.go

```diff
@@ -0,0 +1,164 @@
+package opentracing
+
+import "sync"
+
+// A TraceContextID is the smallest amount of state needed to describe a span's
+// identity within a larger [potentially distributed] trace. The TraceContextID
+// is not intended to encode the span's operation name, timing, or log data,
+// but merely any unique identifiers (etc) needed to contextualize it within a
+// larger trace tree.
+//
+// TraceContextIDs are sufficient to propagate the, well, *context* of a particular
+// trace from process to process.
+type TraceContextID interface {
+	// Create a child context for this TraceContextID, and return both that child's
+	// own TraceContextID as well as any Tags that should be added to the child's
+	// Span.
+	//
+	// The returned TraceContextID type must be the same as the type of the
+	// TraceContextID implementation itself.
+	NewChild() (childCtx TraceContextID, childSpanTags Tags)
+
+	// Serializes the TraceContextID as a printable ASCII string (e.g.,
+	// base64).
+	SerializeASCII() string
+
+	// Serializes the TraceContextID as arbitrary binary data.
+	SerializeBinary() []byte
+}
+
+// A long-lived interface that knows how to create a root TraceContextID and
+// serialize/deserialize any other.
+type TraceContextIDSource interface {
+	// Create a TraceContextID which has no parent (and thus begins its own trace).
+	// A TraceContextIDSource must always return the same type in successive calls
+	// to NewRootTraceContextID().
+	NewRootTraceContextID() TraceContextID
+
+	// Converts the encoded binary data (see
+	// `TraceContextID.SerializeBinary()`) into a TraceContextID of the same
+	// type as returned by NewRootTraceContextID().
+	DeserializeBinaryTraceContextID(encoded []byte) (TraceContextID, error)
+
+	// Converts the encoded ASCII data (see `TraceContextID.SerializeASCII()`)
+	// into a TraceContextID of the same type as returned by
+	// NewRootTraceContextID().
+	DeserializeASCIITraceContextID(encoded string) (TraceContextID, error)
+}
+
+// A `TraceContext` builds off of an implementation-provided TraceContextID and
+// adds a simple string map of "trace tags". The trace tags are special in that
+// they are propagated *in-band*, presumably alongside application data and the
+// `TraceContextID` proper. See the documentation for `SetTraceTag()` for more
+// details and some important caveats.
+//
+// Note that the `TraceContext` is managed internally by the opentracer system;
+// opentracer implementations only need to concern themselves with the
+// `TraceContextID` (which does not know about trace tags).
+type TraceContext struct {
```

Comment:
ok. I don't have a strong preference as well, was just concerned that once we publish the API with a concrete class, it will be hard to change to an interface later. Actually, in my Python version I don't have a 'context ID' abstraction, only Trace Context, since ID and tags are always serialized together, and ID has practically no meaning to anyone by itself, I didn't want to instantiate an extra object.